### ISSUE TITLE
Proposal: DAG image builder (can execute muti-stage build in parallel)

### ISSUE THREAD

**Issue by AkihiroSuda** [moby/moby#32550]
## Proposal

Beginning from 17.05, Docker supports "multi-stage" Dockerfile.

My proposal is to convert such Dockerfile to DAG internally, and execute it in parallel.

![image](https://cloud.githubusercontent.com/assets/9248427/24951034/00b9326c-1fae-11e7-8ade-c9956e3f0693.png)

No change to file format nor UX

POC is available: https://github.com/docker/docker/pull/32402

### Implement generic DAG utility

I made a small generic DAG package: https://github.com/AkihiroSuda/go-dag

```go
import (
    "github.com/AkihiroSuda/go-dag"
    "github.com/AkihiroSuda/go-dag/scheduler"
)

g := &dag.Graph{
	Nodes: []dag.Node{0, 1, 2},
	Edges: []dag.Edge{
		{Depender: 2, Dependee: 0},
		{Depender: 2, Dependee: 1},
	},
}
concurrency := 0
scheduler.Execute(g, concurrency, func(n dag.Node) { buildStage(n) })
```

If this design is correct, I think we can vendor this package (or just copy it to `github.com/docker/docker/pkg/dag`)

### Determine Dockerfile DAG granularity

Alternatively, we could use fine-grained DAG like this, but I'm -1 ATM, because it is likely to cause implementation issue

![image](https://cloud.githubusercontent.com/assets/9248427/24951076/27fc4f12-1fae-11e7-88e8-0b63845e86fd.png)

### Refactor `builder` pkg to create DAG

My previous POC (https://github.com/docker/docker/pull/32402) was implemented in weird way:

- `builder/dockerfile/parser`(unchanged): parses `Dockerfile` text and returns parsed tree structure
- `builder/dockerfile/parallel`: re-parses output from `builder/dockerfile/parser`, and creates DAG

Re-parsing output from  `builder/dockerfile/parser` is likely to cause implementation issue; actually I forgot to implement support for `ARG`.

Rather, we should refactor `builder/dockerfile/parser` to emit DAG directly.

### Investigate other usecases of DAG

https://github.com/docker/docker/pull/32402#issuecomment-292254850

> The DAG based execution engine should be the new core of the builder, not only provide concurrent stage execution but concurrent build jobs, more efficient processing and cache reuse, cache export/import for DAG branches etc. In the same time, this core should be separated from the frontend(Dockerfile) logic to provide more options for declaring the build definition and also provide more extension points for others to reuse this solver.

### Investigate why `docker build` is slow

Storage driver seems being bottleneck, but haven't looked into this deeper

https://github.com/docker/docker/issues/9656
https://github.com/docker/docker/pull/32402#issuecomment-292313684
https://github.com/docker/docker/pull/32402#issuecomment-292888210

cc @tonistiigi @dnephin @aaronlehmann @vdemeester @cpuguy83 @simonferquel @thaJeztah

**Comment by simonferquel** [comment]
I would prefer to have a fine grained DAG. Possibly even more fine grained than a direct equivalent to the dockerfile commands (opening optimizations like auto commit squashing, squashing of multiple ENV / COPY calls, more precise dependency management opening more parallelism etc.)

For exemple if you take
```
FROM img AS foo
<...>
FROM img2
RUN <...>
COPY --from=foo /foo /bar
```
we could run foo and img2 context in parallel until we actually need to COPY from foo.
It would also be very usefull to have this fine grained DAG observable so that we can do a PrePass where we could do Semantic validation, and repo list extraction (very usefull for the upcoming credential on demand streaming)

**Comment by thaJeztah** [comment]
Fine grained DAG could be a continuation on this proposal? I see benefits in having that, but also agree that adding complication at the start may not be wise; happy to hear from others though!

**Comment by duglin** [comment]
While this looks like an interesting optimization, given the newness of some of the features this requires (like the multi-stage build), I'd prefer to wait until we know how well people react and use the existing new build features before we introduce something like this.

**Comment by tonistiigi** [comment]
In addition to my comments in https://github.com/moby/moby/pull/32402#issuecomment-292254850 I wrote a proposal for a direction this could take https://gist.github.com/tonistiigi/059fc72c4630f066d94dafb5e0e70dc6 . Please note that that is only a proposal, there in nobody developing that atm.

I think the correct ways to get started on this would be:
- Fix builder dependency problem https://github.com/moby/moby/issues/32904
- Implement parsing of build stages so they can be preprocessed and analyzed
- Remove linear execution of build stages. Make sure unnecessary stages are skipped.
- Make sure multiple builder invocations don't duplicate workloads.

I think specifically skipping unnecessary stages is a good goal to take before tackling what configuration of parallelization would be optimum. It would also make it possible for https://github.com/moby/moby/issues/32487 to build images for multiple operating systems.

@AkihiroSuda or anyone else, if the above looks good to you, let me know if you would be interested in helping with any of this or #32904 (or actually any builder proposal). We can do a quick design discussion or live design doc for the specific items.

**Comment by AkihiroSuda** [comment]
@tonistiigi

Thank you a lot for working on this, SGTM :+1:.

> Connection with docker/docker project

Now this section would need to be rewritten for "docker product" and "moby project" (cc @shykes @moby/moby-maintainers)

> Sources

With regarding to "git repositories, http archives" sources, do you plan to add more remote sources? (Then it might be good to consider more modularization) Or are they just for compatibility with the current implementation?

> Snapshots

I wonder we can just use containerd snapshot drivers for simplicity.
Non-standard drivers could be implemented as some 3rd party containerd plugin

> Cache import/export

:+1:

Also, it might be good to consider P2P caching here?
(Scope of docker/distribution?)

**Comment by tonistiigi** [comment]
@AkihiroSuda 

> With regarding to "git repositories, http archives" sources, do you plan to add more remote sources? (Then it might be good to consider more modularization) Or are they just for compatibility with the current implementation?

Buildkit itself should be quite open about adding new sources. `docker build` would probably ship with a more stricter set. Buildkit in its core is a library so if you are thinking of having pluggable sources then that would just need to be implemented by a tool that includes it.

> I wonder we can just use containerd snapshot drivers for simplicity.
Non-standard drivers could be implemented as some 3rd party containerd plugin

We definitely one to rely on containerd snapshot drivers for the actual implementation. One of the issues is that for `docker build` use case we need to share the snapshots with other components(like `docker images`). So currently I see it as a thin wrapper around containerd drivers that provides reference counting capabilities.

> Also, it might be good to consider P2P caching here?

Agree, but one problem at a time.

I think I'll open up a new issue with that proposal even if the code for it would not end up in this repo so we don't have to share the gist. We can discuss there if we are ready to open a new repo etc.

**Comment by AkihiroSuda** [comment]
Should we close this issue for housekeeping purpose?

Now we have buildkit repo: https://github.com/moby/buildkit

**Comment by dnephin** [comment]
Let's keep it open to track this issue in this repo.

**Comment by AkihiroSuda** [comment]
ok

**Comment by AkihiroSuda** [comment]
Now BuildKit supports converting Dockerfile to LLB DAG :tada:
https://github.com/moby/buildkit/pull/106

**Comment by vdemeester** [comment]
@tonistiigi @AkihiroSuda should we close this one as `buildkit` supports it (mainly to not forget closing this one once buildkit is integrated in moby).

**Comment by vdemeester** [comment]
@AkihiroSuda yes :+1:

**Comment by AkihiroSuda** [comment]
Let's close this one as #37151 has been merged now 🎉
